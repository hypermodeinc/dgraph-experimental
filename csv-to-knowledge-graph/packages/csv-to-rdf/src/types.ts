export interface RdfMap {
  [key: string]: string | string[];
}

export interface XidMap {
  [key: string]: string;
}

export interface DfRow {
  [key: string]: any;
  LINE_NUMBER?: number;
}

export interface OutputOptions {
  onChunk?: (chunk: string) => void;
  onComplete?: (result: string) => void;
  onError?: (error: Error) => void;
}
