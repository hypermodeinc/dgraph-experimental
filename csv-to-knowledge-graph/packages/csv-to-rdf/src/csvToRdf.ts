import Papa from "papaparse";
import { transformDataFrame } from "./rdfLib";
import type { XidMap, DfRow, RdfMap } from "./types";
import {
  parserLogger,
  processingLogger,
  conversionLogger,
  templateLogger,
} from "./debug";

interface CsvToRdfOptions {
  template: string;
  chunkSize?: number;
  xidmap?: XidMap;
  onProgress?: (progress: number) => void;
  normalizeHeaders?: boolean; // New option to control header normalization
}

export class CsvToRdf {
  private template: string;
  private chunkSize: number;
  private xidmap: XidMap;
  private rdfBuffer: string[];
  private completed: boolean = false;
  private onProgress?: (progress: number) => void;
  private normalizeHeaders: boolean;
  private headerMapping: Map<string, string> = new Map();

  constructor(options: CsvToRdfOptions) {
    this.template = options.template;
    this.chunkSize = options.chunkSize || 1000;
    this.xidmap = options.xidmap || {};
    this.rdfBuffer = [];
    this.onProgress = options.onProgress;
    this.normalizeHeaders = options.normalizeHeaders !== false; // Default to true

    processingLogger.info(
      "CsvToRdf instance created with chunk size: %d",
      this.chunkSize,
    );
    templateLogger.info("Template length: %d characters", this.template.length);
    templateLogger.info(
      "Header normalization: %s",
      this.normalizeHeaders ? "enabled" : "disabled",
    );
  }

  /**
   * Normalize header by removing quotes and trimming whitespace
   */
  private normalizeHeader(header: string): string {
    if (!this.normalizeHeaders) return header;

    // Remove outer quotes (both single and double)
    let normalized = header.replace(/^["']|["']$/g, "").trim();

    // Handle escaped quotes inside headers
    normalized = normalized.replace(/""/g, '"').replace(/''/g, "'");

    templateLogger.info('Header normalized: "%s" -> "%s"', header, normalized);
    return normalized;
  }

  /**
   * Create mapping between original and normalized headers
   */
  private createHeaderMapping(headers: string[]): void {
    this.headerMapping.clear();

    headers.forEach((header) => {
      const normalized = this.normalizeHeader(header);
      this.headerMapping.set(header, normalized);
      // Also map normalized to normalized for convenience
      this.headerMapping.set(normalized, normalized);
    });

    templateLogger.info(
      "Created header mapping for %d headers",
      headers.length,
    );
    templateLogger.info(
      "Header mapping: %o",
      Object.fromEntries(this.headerMapping),
    );
  }

  /**
   * Transform headers in parsed data rows
   */
  private transformRowHeaders(rows: any[]): DfRow[] {
    if (!this.normalizeHeaders || this.headerMapping.size === 0) {
      return rows as DfRow[];
    }

    return rows.map((row) => {
      const transformedRow: DfRow = {};

      Object.keys(row).forEach((originalHeader) => {
        const normalizedHeader =
          this.headerMapping.get(originalHeader) || originalHeader;
        transformedRow[normalizedHeader] = row[originalHeader];
      });

      return transformedRow;
    });
  }

  async processCSVString(csvData: string): Promise<string> {
    processingLogger.info(
      "Processing CSV string of length: %d",
      csvData.length,
    );

    return new Promise((resolve, reject) => {
      let totalRows = 0;
      let headers: string[] = [];

      parserLogger.info(
        "Starting initial parse to count rows and extract headers",
      );
      Papa.parse(csvData, {
        header: true,
        dynamicTyping: true,
        skipEmptyLines: true,
        transformHeader: (header: string) => {
          const normalized = this.normalizeHeader(header);
          templateLogger.info(
            'Transform header: "%s" -> "%s"',
            header,
            normalized,
          );
          return normalized;
        },
        step: (results: any) => {
          if (totalRows === 0) {
            // Capture headers from first row
            headers = results.meta.fields || [];
            this.createHeaderMapping(headers);
            parserLogger.info("Extracted headers: %o", headers);
          }
          totalRows++;
        },
        complete: () => {
          parserLogger.info("Counted %d total rows in CSV", totalRows);
          this.processWithProgress(csvData, totalRows, resolve, reject);
        },
        error: (error: Error) => {
          parserLogger.error("Error counting rows: %o", error);
          reject(new Error(`Error counting rows: ${error.message}`));
        },
      });
    });
  }

  private processWithProgress(
    csvData: string,
    totalRows: number,
    resolve: (value: string) => void,
    reject: (reason: Error) => void,
  ) {
    let processedRows = 0;

    processingLogger.info(
      "Starting CSV processing with %d total rows",
      totalRows,
    );

    Papa.parse(csvData, {
      header: true,
      dynamicTyping: true,
      skipEmptyLines: true,
      chunkSize: this.chunkSize,
      transformHeader: (header: string) => {
        return this.normalizeHeader(header);
      },
      chunk: (results: any) => {
        processingLogger.info(
          "Processing chunk with %d rows",
          results.data.length,
        );

        // Transform headers in the chunk data
        const transformedRows = this.transformRowHeaders(results.data);

        const rdfOutput = this.transformChunk(transformedRows);
        if (rdfOutput) {
          this.rdfBuffer.push(rdfOutput);
          processingLogger.info(
            "Generated %d bytes of RDF data",
            rdfOutput.length,
          );
        } else {
          processingLogger.warning("No RDF data generated from chunk");
        }

        processedRows += results.data.length;
        const progress = Math.min(
          100,
          Math.floor((processedRows / totalRows) * 100),
        );
        processingLogger.info(
          "Progress: %d%% (%d/%d rows)",
          progress,
          processedRows,
          totalRows,
        );

        if (this.onProgress) {
          this.onProgress(progress);
        }
      },
      complete: () => {
        this.completed = true;
        if (this.onProgress) {
          this.onProgress(100);
        }
        processingLogger.info(
          "CSV processing completed, generated %d chunks of RDF data",
          this.rdfBuffer.length,
        );
        resolve(this.rdfBuffer.join(""));
      },
      error: (error: Error) => {
        processingLogger.error("Error processing CSV: %o", error);
        reject(new Error(`Error processing CSV: ${error.message}`));
      },
    });
  }

  async processCSVFile(file: File): Promise<string> {
    processingLogger.info(
      "Processing CSV file: %s (%d bytes)",
      file.name,
      file.size,
    );

    return new Promise((resolve, reject) => {
      let totalRows = 0;
      let headers: string[] = [];

      parserLogger.info(
        "Starting initial parse to count rows in file: %s",
        file.name,
      );
      Papa.parse(file, {
        header: true,
        dynamicTyping: true,
        skipEmptyLines: true,
        transformHeader: (header: string) => {
          return this.normalizeHeader(header);
        },
        step: (results: any) => {
          if (totalRows === 0) {
            // Capture headers from first row
            headers = results.meta.fields || [];
            this.createHeaderMapping(headers);
            parserLogger.info("Extracted headers from file: %o", headers);
          }
          totalRows++;
        },
        complete: () => {
          parserLogger.info("Counted %d total rows in file", totalRows);
          this.processFileWithProgress(file, totalRows, resolve, reject);
        },
        error: (error) => {
          parserLogger.error("Error counting rows in file: %o", error);
          reject(new Error(`Error counting rows: ${error.message}`));
        },
      });
    });
  }

  private processFileWithProgress(
    file: File,
    totalRows: number,
    resolve: (value: string) => void,
    reject: (reason: Error) => void,
  ) {
    let processedRows = 0;

    processingLogger.info(
      "Starting file processing with %d total rows",
      totalRows,
    );

    Papa.parse(file, {
      header: true,
      dynamicTyping: true,
      skipEmptyLines: true,
      chunkSize: this.chunkSize,
      transformHeader: (header: string) => {
        return this.normalizeHeader(header);
      },
      chunk: (results) => {
        processingLogger.info(
          "Processing file chunk with %d rows",
          results.data.length,
        );

        // Transform headers in the chunk data
        const transformedRows = this.transformRowHeaders(results.data);

        const rdfOutput = this.transformChunk(transformedRows);
        if (rdfOutput) {
          this.rdfBuffer.push(rdfOutput);
          processingLogger.info(
            "Generated %d bytes of RDF data from file chunk",
            rdfOutput.length,
          );
        } else {
          processingLogger.warning("No RDF data generated from file chunk");
        }

        processedRows += results.data.length;
        const progress = Math.min(
          100,
          Math.floor((processedRows / totalRows) * 100),
        );
        processingLogger.info(
          "File processing progress: %d%% (%d/%d rows)",
          progress,
          processedRows,
          totalRows,
        );

        if (this.onProgress) {
          this.onProgress(progress);
        }
      },
      complete: () => {
        this.completed = true;
        if (this.onProgress) {
          this.onProgress(100);
        }
        processingLogger.info(
          "File processing completed, generated %d chunks of RDF data",
          this.rdfBuffer.length,
        );
        resolve(this.rdfBuffer.join(""));
      },
      error: (error) => {
        processingLogger.error("Error processing file: %o", error);
        reject(new Error(`Error processing CSV: ${error.message}`));
      },
    });
  }

  private transformChunk(rows: DfRow[]): string | null {
    conversionLogger.info("Transforming chunk with %d rows", rows.length);

    const rowsWithLineNumbers = rows.map((row, index) => ({
      ...row,
      LINE_NUMBER: index,
    }));

    try {
      const rdfMap = transformDataFrame(rowsWithLineNumbers, this.template);
      const triplesCount = Object.keys(rdfMap).length;
      conversionLogger.info("Generated %d triples from chunk", triplesCount);

      const result = this.rdfMapToString(rdfMap);
      return result;
    } catch (error) {
      conversionLogger.error("Error transforming chunk to RDF: %o", error);
      return null;
    }
  }

  private rdfMapToString(rdfMap: RdfMap): string {
    let result = "";
    let triplesCount = 0;

    for (const [subject, value] of Object.entries(rdfMap)) {
      if (Array.isArray(value)) {
        for (const v of value) {
          result += `${subject} ${v} .\n`;
          triplesCount++;
        }
      } else {
        result += `${subject} ${value} .\n`;
        triplesCount++;
      }
    }

    conversionLogger.info(
      "Converted RDF map to string with %d triples",
      triplesCount,
    );
    return result;
  }

  isCompleted(): boolean {
    return this.completed;
  }

  getXidMap(): XidMap {
    return this.xidmap;
  }

  /**
   * Get the header mapping (original -> normalized)
   */
  getHeaderMapping(): Map<string, string> {
    return new Map(this.headerMapping);
  }

  /**
   * Get normalized headers
   */
  getNormalizedHeaders(): string[] {
    return Array.from(new Set(this.headerMapping.values()));
  }
}

export function createCsvToRdf(
  template: string,
  options?: Omit<CsvToRdfOptions, "template">,
): CsvToRdf {
  processingLogger.info(
    "Creating CsvToRdf instance with template of length %d",
    template.length,
  );
  return new CsvToRdf({ template, ...options });
}
