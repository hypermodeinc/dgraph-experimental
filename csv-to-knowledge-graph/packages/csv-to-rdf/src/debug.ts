import { createLogger } from "@hypermode/csvkit-utils";

const logger = createLogger("csv-to-rdf");

export const parserLogger = logger("parser");
export const templateLogger = logger("template");
export const processingLogger = logger("processing");
export const conversionLogger = logger("conversion");
export const debugLogger = logger("debug");
