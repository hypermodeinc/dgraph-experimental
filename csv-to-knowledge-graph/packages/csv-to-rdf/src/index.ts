export { CsvToRdf, createCsvToRdf } from "./csvToRdf";
export {
  processBatchToRDF,
  createBatchRdfProcessor,
  analyzeCSVHeaders,
} from "./batchProcessor";
export { dfToRdfMap, transformDataFrame } from "./rdfLib";
export type { RdfMap, XidMap } from "./types";
export type { CSVBatchFile, BatchRdfOptions } from "./batchProcessor";
