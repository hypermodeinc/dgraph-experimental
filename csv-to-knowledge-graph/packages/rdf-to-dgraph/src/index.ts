export { RdfToDgraph, importRdfToDgraph } from "./rdfToDgraph";
export {
  parseDgraphConnectionString,
  connectionOptionsToDgraphCredentials,
  parseDgraphUrl,
} from "./dgraphUrl";
export * from "./types";
