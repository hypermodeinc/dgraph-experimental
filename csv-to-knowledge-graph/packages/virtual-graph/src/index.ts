export { Graph, GraphChanges, Node, Edge } from "./types";
export { mergeGraphChanges } from "./mergeGraphChanges";
export {
  parseGraphData,
  validateGraph,
  getConnectedNodes,
  getGraphStats,
  findShortestPath,
  calculateNodePositions,
} from "./utils";
export { assignNodeLevels, createHierarchy } from "./graphHierarchy";

export type {
  GraphData,
  EdgeIdentifier,
  NodeModification,
  PropertyAddition,
  PropertyDeletion,
} from "./types";
