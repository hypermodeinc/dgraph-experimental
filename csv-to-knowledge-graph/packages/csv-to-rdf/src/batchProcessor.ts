import { createCsvToRdf } from "./csvToRdf";
import { processingLogger } from "./debug";

export interface CSVBatchFile {
  id: string;
  name: string;
  content: string;
}

export interface BatchRdfOptions {
  onProgress?: (progress: number) => void;
  onStatusChange?: (status: string) => void;
  normalizeHeaders?: boolean;
}

export async function processBatchToRDF(
  files: CSVBatchFile[],
  template: string,
  options: BatchRdfOptions = {},
): Promise<string> {
  processingLogger.info(
    "Starting batch RDF processing with %d files",
    files.length,
  );
  if (files.length === 0) {
    processingLogger.warning("No files provided for batch processing");
    return "";
  }

  if (!template) {
    processingLogger.warning("No template provided for batch processing");
    return "";
  }

  // Optional callbacks
  const { onProgress, onStatusChange, normalizeHeaders = true } = options;

  try {
    onStatusChange?.("Analyzing template and CSV files...");
    onProgress?.(10);

    // Step 1: First pass - Generate entity definitions without relationships
    processingLogger.info("First pass - generating entity definitions");
    onStatusChange?.("Generating entity definitions from CSV files...");

    // Extract entity template lines (those that don't define relationships between entities)
    const templateLines = template.split("\n");
    const entityTemplateLines = templateLines
      .filter((line) => {
        // Keep lines that are NOT relationships between entities
        // Relationships typically have a pattern where both subject and object are entity references like <_:Entity>
        return !line.includes("> <_:") && !line.match(/>\s+<[A-Z_]+>\s+<_:/);
      })
      .join("\n");

    // First pass - generate all entity definitions
    let entityDefinitions = "";
    let processedFiles = 0;

    for (const file of files) {
      if (!file.content) continue;

      processingLogger.info(
        "Processing file %s for entity definitions",
        file.name,
      );

      const converter = createCsvToRdf(entityTemplateLines, {
        normalizeHeaders, // Pass through header normalization option
        onProgress: (fileProgress) => {
          // Map progress for first pass (0-40%)
          const fileWeight = 1 / files.length;
          const baseProgress = 10 + processedFiles * fileWeight * 30;
          const fileProgress1 = fileWeight * (fileProgress / 100) * 30;
          onProgress?.(Math.floor(baseProgress + fileProgress1));
        },
      });

      // Process this file for entities
      onStatusChange?.(`Generating entities from ${file.name}...`);
      const fileEntityData = await converter.processCSVString(file.content);
      entityDefinitions += fileEntityData;

      const headerMapping = converter.getHeaderMapping();
      processingLogger.info(
        "Header mapping for %s: %o",
        file.name,
        Object.fromEntries(headerMapping),
      );

      processedFiles++;
    }

    // Step 2: Second pass - Extract only relationship triples
    processingLogger.info("Second pass - generating relationship triples");
    onStatusChange?.("Generating relationships between entities...");
    onProgress?.(45);

    // Extract relationship template lines
    const relationshipTemplateLines = templateLines
      .filter((line) => {
        // Keep only lines that define relationships between entities
        return (
          (line.includes("> <_:") || line.match(/>\s+<[A-Z_]+>\s+<_:/)) &&
          !line.includes("dgraph.type")
        );
      })
      .join("\n");

    // Second pass - generate all relationship triples
    let relationshipTriples = "";
    processedFiles = 0;

    for (const file of files) {
      if (!file.content) continue;

      processingLogger.info("Processing file %s for relationships", file.name);

      const converter = createCsvToRdf(relationshipTemplateLines, {
        normalizeHeaders,
        onProgress: (fileProgress) => {
          // Map progress for second pass (45-80%)
          const fileWeight = 1 / files.length;
          const baseProgress = 45 + processedFiles * fileWeight * 35;
          const fileProgress2 = fileWeight * (fileProgress / 100) * 35;
          onProgress?.(Math.floor(baseProgress + fileProgress2));
        },
      });

      // Process this file for relationships
      onStatusChange?.(`Generating relationships from ${file.name}...`);
      try {
        const fileRelationshipData = await converter.processCSVString(
          file.content,
        );
        relationshipTriples += fileRelationshipData;
      } catch (err) {
        // If relationship processing fails for a file, log but continue
        processingLogger.warning(
          "Error processing relationships for %s: %o",
          file.name,
          err,
        );
      }

      processedFiles++;
    }

    // Step 3: Clean and merge the data
    onStatusChange?.("Finalizing RDF data...");
    onProgress?.(85);

    // Function to clean RDF data
    const cleanRdfData = (rdfData: string): string => {
      return rdfData
        .split("\n")
        .map((line) => line.trim())
        .filter((line) => {
          // Skip empty lines
          if (!line) return false;
          // Skip lines with empty quotes
          if (line.includes('""')) return false;
          // Skip lines with unsubstituted placeholders
          if (line.includes("[") && line.includes("]")) return false;
          return true;
        })
        .join("\n");
    };

    // Clean both entity definitions and relationship triples
    const cleanedEntityDefinitions = cleanRdfData(entityDefinitions);
    const cleanedRelationshipTriples = cleanRdfData(relationshipTriples);

    // Combine - entities first, then relationships
    const combinedRdfData =
      cleanedEntityDefinitions + "\n" + cleanedRelationshipTriples;

    // Step 4: Remove duplicate triples
    const uniqueLines = new Set(
      combinedRdfData.split("\n").filter((line) => line.trim()),
    );
    const dedupedRdfData = Array.from(uniqueLines).join("\n");

    processingLogger.info(
      "Batch processing complete. Generated %d unique RDF triples",
      uniqueLines.size,
    );
    onStatusChange?.("RDF generation complete");
    onProgress?.(100);

    return dedupedRdfData;
  } catch (error) {
    processingLogger.error("Error in batch processing: %o", error);
    throw error;
  }
}

/**
 * Process a batch of CSV files with a shared RDF template
 * for use in scenarios that require manual setup like browser environments
 */
export function createBatchRdfProcessor() {
  return {
    processBatchToRDF,
  };
}

/**
 * Utility function to analyze headers across multiple CSV files
 */
export async function analyzeCSVHeaders(files: CSVBatchFile[]): Promise<{
  fileHeaders: Record<string, string[]>;
  commonHeaders: string[];
  allHeaders: string[];
  headerVariations: Record<string, string[]>;
}> {
  const fileHeaders: Record<string, string[]> = {};
  const allHeadersSet = new Set<string>();
  const headerVariations: Record<string, string[]> = {};

  for (const file of files) {
    if (!file.content) continue;

    const converter = createCsvToRdf("", {
      // Empty template for analysis
      normalizeHeaders: true,
    });

    try {
      // Process just enough to get headers
      await converter.processCSVString(
        file.content.split("\n").slice(0, 2).join("\n"),
      );

      const normalizedHeaders = converter.getNormalizedHeaders();
      const headerMapping = converter.getHeaderMapping();

      fileHeaders[file.name] = normalizedHeaders;

      // Track all unique headers
      normalizedHeaders.forEach((header) => allHeadersSet.add(header));

      // Track header variations
      headerMapping.forEach((normalized, original) => {
        if (normalized !== original) {
          if (!headerVariations[normalized]) {
            headerVariations[normalized] = [];
          }
          if (!headerVariations[normalized].includes(original)) {
            headerVariations[normalized].push(original);
          }
        }
      });
    } catch (error) {
      processingLogger.warning(
        "Could not analyze headers for file %s: %o",
        file.name,
        error,
      );
      fileHeaders[file.name] = [];
    }
  }

  const allHeaders = Array.from(allHeadersSet);

  // Find common headers across all files
  const commonHeaders = allHeaders.filter((header) =>
    Object.values(fileHeaders).every((headers) => headers.includes(header)),
  );

  processingLogger.info(
    "Header analysis complete: %d files, %d unique headers, %d common headers",
    files.length,
    allHeaders.length,
    commonHeaders.length,
  );

  return {
    fileHeaders,
    commonHeaders,
    allHeaders,
    headerVariations,
  };
}
