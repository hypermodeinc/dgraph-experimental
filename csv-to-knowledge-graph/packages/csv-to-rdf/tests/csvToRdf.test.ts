import { CsvToRdf } from "../src/csvToRdf";

jest.mock("papaparse", () => {
  const originalPapa = jest.requireActual("papaparse");

  return {
    ...originalPapa,
    parse: (input: any, config: any) => {
      if (input && typeof input.stream === "function") {
        const content = input.content || "";
        return originalPapa.parse(content, config);
      }
      return originalPapa.parse(input, config);
    },
  };
});

const mockTemplate = `
<_:School_[School.ID]> <dgraph.type> "School" .
<_:School_[School.ID]> <School.name> "[School.Name]" .
<_:Project_[Project.ID]> <dgraph.type> "Project" .
<_:Project_[Project.ID]> <Project.title> "[Project.Title]" .
<_:Project_[Project.ID]> <Project.school> <_:School_[School.ID]> .
`;

const quotedHeaderTemplate = `
<_:Project_[Project.ID]> <dgraph.type> "Project" .
<_:Project_[Project.ID]> <Project.id> "[Project.ID]" .
<_:Donation_[Donation.ID]> <dgraph.type> "Donation" .
<_:Donation_[Donation.ID]> <Donation.id> "[Donation.ID]" .
<_:Donation_[Donation.ID]> <Donation.amount> "[Donation.Amount]" .
<_:Donor_[Donor.ID]> <dgraph.type> "Donor" .
<_:Donor_[Donor.ID]> <Donor.id> "[Donor.ID]" .
<_:Donation_[Donation.ID]> <Donation.project> <_:Project_[Project.ID]> .
<_:Donation_[Donation.ID]> <Donation.donor> <_:Donor_[Donor.ID]> .
`;

function createMockCsvString(data: string[][]): string {
  return data.map((row) => row.join(",")).join("\n");
}

class MockFile {
  content: string;
  name: string;
  type: string;

  constructor(content: string, filename: string, options: any = {}) {
    this.content = content;
    this.name = filename;
    this.type = options.type || "text/csv";
  }

  // Minimal implementation to be recognized as File object
  stream() {
    return {};
  }
}

describe("CsvToRdf", () => {
  const mockCsvData = [
    ["School.ID", "School.Name", "Project.ID", "Project.Title"],
    ["SCHOOL-001", "Lincoln High School", "PROJ-001", "Math Competition Gear"],
    [
      "SCHOOL-002",
      "Roosevelt Elementary",
      "PROJ-002",
      "Science Fair Materials",
    ],
  ];

  const quotedHeaderCsvData = [
    ['"Project.ID"', '"Donation.ID"', '"Donor.ID"', '"Donation.Amount"'],
    ['"PROJ001"', '"DON001"', '"DONOR001"', '"100.50"'],
    ['"PROJ002"', '"DON002"', '"DONOR002"', '"25.00"'],
  ];

  test("basic snapshot test", async () => {
    const csvString = createMockCsvString(mockCsvData);
    const converter = new CsvToRdf({ template: mockTemplate });

    const result = await converter.processCSVString(csvString);
    expect(result).toMatchSnapshot();
  });

  test("processes with progress tracking", async () => {
    const csvString = createMockCsvString(mockCsvData);
    const progressValues: number[] = [];

    const converter = new CsvToRdf({
      template: mockTemplate,
      onProgress: (progress) => {
        progressValues.push(progress);
      },
    });

    const result = await converter.processCSVString(csvString);

    expect(result).toBeTruthy();
    expect(progressValues.length).toBeGreaterThan(0);
    expect(progressValues[progressValues.length - 1]).toBe(100);
  });

  test("processes CSV file correctly", async () => {
    const csvString = createMockCsvString(mockCsvData);
    const file = new MockFile(csvString, "test.csv", { type: "text/csv" });

    const converter = new CsvToRdf({ template: mockTemplate });
    const result = await converter.processCSVFile(file as any);

    expect(result).toMatchSnapshot();
  });

  test("handles empty CSV correctly", async () => {
    const emptyData = [["Column1", "Column2"]];
    const csvString = createMockCsvString(emptyData);

    const converter = new CsvToRdf({ template: mockTemplate });
    const result = await converter.processCSVString(csvString);

    expect(result).toBe("");
  });

  test("handles large CSV with chunking", async () => {
    const largeData = [
      ["School.ID", "School.Name", "Project.ID", "Project.Title"],
    ];
    for (let i = 1; i <= 100; i++) {
      largeData.push([
        `SCHOOL-${i}`,
        `School ${i}`,
        `PROJ-${i}`,
        `Project ${i}`,
      ]);
    }

    const csvString = createMockCsvString(largeData);
    const converter = new CsvToRdf({
      template: mockTemplate,
      chunkSize: 10,
    });

    const result = await converter.processCSVString(csvString);

    const lines = result.split("\n").filter((line) => line.trim());
    expect(lines.length).toBeGreaterThan(400);
  });

  // New tests for quoted headers
  test("handles quoted headers correctly", async () => {
    const csvString = createMockCsvString(quotedHeaderCsvData);
    const converter = new CsvToRdf({
      template: quotedHeaderTemplate,
      normalizeHeaders: true,
    });

    const result = await converter.processCSVString(csvString);

    // Should contain processed data without quoted header issues
    expect(result).toContain("Project.id");
    expect(result).toContain("Donation.id");
    expect(result).toContain("Donor.id");
    expect(result).toContain('"PROJ001"');
    expect(result).toContain('"DON001"');
    expect(result).toContain('"DONOR001"');
  });

  test("header mapping works correctly", async () => {
    const csvString = createMockCsvString(quotedHeaderCsvData);
    const converter = new CsvToRdf({
      template: quotedHeaderTemplate,
      normalizeHeaders: true,
    });

    await converter.processCSVString(csvString);

    const headerMapping = converter.getHeaderMapping();
    const normalizedHeaders = converter.getNormalizedHeaders();

    expect(headerMapping.size).toBeGreaterThan(0);
    expect(normalizedHeaders).toContain("Project.ID");
    expect(normalizedHeaders).toContain("Donation.ID");
    expect(normalizedHeaders).toContain("Donor.ID");
    expect(normalizedHeaders).toContain("Donation.Amount");
  });

  test("can disable header normalization", async () => {
    const csvString = createMockCsvString(quotedHeaderCsvData);
    const converter = new CsvToRdf({
      template: quotedHeaderTemplate,
      normalizeHeaders: false,
    });

    await converter.processCSVString(csvString);

    const headerMapping = converter.getHeaderMapping();

    // With normalization disabled, mapping should be minimal
    expect(headerMapping.size).toBeLessThanOrEqual(
      quotedHeaderCsvData[0].length * 2,
    );
  });

  test("handles mixed quote styles", async () => {
    const mixedQuoteData = [
      ['"Project.ID"', "'Donation.ID'", "Donor.ID", '"Donation.Amount"'],
      ['"PROJ001"', "'DON001'", "DONOR001", '"100.50"'],
      ['"PROJ002"', "'DON002'", "DONOR002", '"25.00"'],
    ];

    const csvString = createMockCsvString(mixedQuoteData);
    const converter = new CsvToRdf({
      template: quotedHeaderTemplate,
      normalizeHeaders: true,
    });

    const result = await converter.processCSVString(csvString);

    expect(result).toContain("Project.id");
    expect(result).toContain("Donation.id");
    expect(result).toContain("Donor.id");
  });

  test("handles escaped quotes in headers", async () => {
    const escapedQuoteData = [
      ['"Project ""Special"" ID"', '"Donation.ID"', '"Donor.ID"'],
      ['"PROJ001"', '"DON001"', '"DONOR001"'],
    ];

    const templateWithEscaped = `
<_:Project_[Project "Special" ID]> <dgraph.type> "Project" .
<_:Donation_[Donation.ID]> <dgraph.type> "Donation" .
<_:Donor_[Donor.ID]> <dgraph.type> "Donor" .
`;

    const csvString = createMockCsvString(escapedQuoteData);
    const converter = new CsvToRdf({
      template: templateWithEscaped,
      normalizeHeaders: true,
    });

    const result = await converter.processCSVString(csvString);

    // Should handle escaped quotes correctly
    expect(result).toBeTruthy();

    const normalizedHeaders = converter.getNormalizedHeaders();
    expect(normalizedHeaders).toContain('Project "Special" ID');
  });

  test("progress tracking works with quoted headers", async () => {
    const csvString = createMockCsvString(quotedHeaderCsvData);
    const progressValues: number[] = [];

    const converter = new CsvToRdf({
      template: quotedHeaderTemplate,
      normalizeHeaders: true,
      onProgress: (progress) => {
        progressValues.push(progress);
      },
    });

    const result = await converter.processCSVString(csvString);

    expect(result).toBeTruthy();
    expect(progressValues.length).toBeGreaterThan(0);
    expect(progressValues[progressValues.length - 1]).toBe(100);
  });
});
