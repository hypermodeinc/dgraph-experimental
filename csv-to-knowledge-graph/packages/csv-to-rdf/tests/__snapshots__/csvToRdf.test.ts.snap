// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvToRdf basic snapshot test 1`] = `
"<_:School_SCHOOL_001> <dgraph.type> "School" .
<_:School_SCHOOL_001> <School.name> "Lincoln High School" .
<_:Project_PROJ_001> <dgraph.type> "Project" .
<_:Project_PROJ_001> <Project.title> "Math Competition Gear" .
<_:Project_PROJ_001> <Project.school> <_:School_SCHOOL_001> .
<_:School_SCHOOL_002> <dgraph.type> "School" .
<_:School_SCHOOL_002> <School.name> "Roosevelt Elementary" .
<_:Project_PROJ_002> <dgraph.type> "Project" .
<_:Project_PROJ_002> <Project.title> "Science Fair Materials" .
<_:Project_PROJ_002> <Project.school> <_:School_SCHOOL_002> .
"
`;

exports[`CsvToRdf processes CSV file correctly 1`] = `
"<_:School_SCHOOL_001> <dgraph.type> "School" .
<_:School_SCHOOL_001> <School.name> "Lincoln High School" .
<_:Project_PROJ_001> <dgraph.type> "Project" .
<_:Project_PROJ_001> <Project.title> "Math Competition Gear" .
<_:Project_PROJ_001> <Project.school> <_:School_SCHOOL_001> .
<_:School_SCHOOL_002> <dgraph.type> "School" .
<_:School_SCHOOL_002> <School.name> "Roosevelt Elementary" .
<_:Project_PROJ_002> <dgraph.type> "Project" .
<_:Project_PROJ_002> <Project.title> "Science Fair Materials" .
<_:Project_PROJ_002> <Project.school> <_:School_SCHOOL_002> .
"
`;
